Documentation/proposals/202305-scrapeclasses.md

```diff
@@ -0,0 +1,223 @@
+The downside is obviously the loss of a great feature of the Prometheus Operator. The monitor resources make it possible
+to compose the scrape configurations in a Kubernetes way.
+
+## Goals
+
+- Allow for the administrator to define a named, reusable scrape configuration snippet, including unsafe elements such as file references.
+- Allow for a user to select a configuration snippet by name in their probe/podmonitor/servicemonitor endpoint configuration.
+- Avoid giving the user the ability to use arbirary files within the Prometheus pod.
+
+### Audience
+
+- Users who serve Prometheus as a service and want to give their customers autonomy in defining monitors.
+- Users who want to scrape pods within an Istio mesh.
+
+## How
+
+The proposed solution is to introduce a notion of a *scrape class*, akin to a Kubernetes [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/).
+A scrape class defines TLS settings (and possibly other settings in future, e.g. sensitive authorization settings) to be applied to all scrape configs of that class.
+
+One scrape class may be designated as the default class, in which case that class is applied to any scrape config that doesn't specify a value for `scrapeClass`.
+
+When defining a podmonitor/servicemonitor/probe/scrapeconfig, a user may assign a scrape class via the `scrapeClass` field.
+For monitors, each endpoint may be assigned a scrape class.
+
+Class names are assumed to be installation-specific. In practice, some common class names like `istio-mtls` are likely to emerge.
+
+### Prometheus Resource
+
+It is proposed that the `Prometheus` and `PrometheusAgent` resources contain a new section for defining scrape classes.
+
+The rationale for defining scrape classes inline is that, in practice, the TLS file paths are closely related to the `volumeMounts`
+of the `Prometheus` spec. An alternative is outlined later, of factoring the class definitions into a separate resource.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+spec:
+  # define scrape classes for use by the monitors
+  # one class may be designated as the default class
+  scrapeClasses:
+    - name: istio-mtls
+      default: true
+      tlsConfig:
+        caFile: "/etc/istio-certs/root-cert.pem"
+        certFile: "/etc/istio-certs/cert-chain.pem"
+        keyFile: "/etc/istio-certs/key.pem"
+        insecureSkipVerify: true
+
+  # mount the certs from the istio sidecar (shown here for illustration purposes)
+  volumeMounts:
+    - name: istio-certs
+      mountPath: "/etc/istio-certs/"
+```
+
+Any object references in the scrape class definition are assumed to refer to objects in the namespace of the `Prometheus` object.
+
+### PodMonitor Resource
+
+Allow the user to select a scrape class for each endpoint.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+spec:
+  podMetricsEndpoints:
+  - port: http
+    path: /metrics
+    scrapeClass: istio-mtls
```

Comment:
Why per endpoint and not per object? I am curious since many `*Class` objects (`runtimeClass`, `storageClass`, `ingressClass`) in kubernetes are applied on the whole child object and not just part of the configuration. Since there is no restriction on how many PodMonitors there can be, shouldn't prometheus-operator follow the same principle as kubernetes?

Reply:
I agree with you @paulfantom

Reply:
I missed this comment but I agree that it makes more sense to operate at the spec level than at the endpoint level.